### PR TITLE
OpenAPI generator upgrade (7.0.0) + Python `lakefs-sdk`

### LINKED ISSUE #6425: [1.0] Forwards-compatible Python clients
Verify that the OpenAPI template used for Python generates a forwards-compatible Python API.  If it isn't, pick or modify (or write...) an OpenAPI template that will generate such an API.

Most importantly: it must be possible to add new optional parameters (to body, query, or even headers) _without_ breaking existing code.

One way to do this is to _force_ keyword arguments on the generated API for optional arguments.  If we do this, need to see how other projects prevent people from using them in positional locations.

Another way is to have the Python language API always take a single "Request" parameter and return a single "Response" parameter.  This is very common: the generated Go client, gRPC (so all Google SDKs), and AWS are all examples.

Out of scope:

* "Wrapper" functions that improve developer experience. For instance upload/download managers.

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
Closes #6425

## Change Description

This PR doesn't actually close the issue, becuase I couldn't find true forward compatibility issues with the current SDK, but it's preferred to upgrade the OpenAPI generator as it's pretty old (Oct 24, 2021)...
This will not break the current behavior and examples from the docs because it's not published as the `lakefs-client` project but as a new `lakefs-sdk` project.

### CODE REVIEW

File: .github/workflows/python-api-client.yaml [hunk 1]
```diff
@@ -28,14 +28,22 @@ jobs:
     - name: Python build and make package
       run: make package-python PACKAGE_VERSION=${{ steps.version.outputs.tag }}
 
-    - name: Python publish package
+    - name: Python client publish package
```
Comment (arielshaqed) [inline .github/workflows/python-api-client.yaml:31]:
```suggestion
    - name: Python SDK publish package
```
?

File: Makefile [hunk 2]
```diff
@@ -12,9 +12,14 @@ PROTOC=$(DOCKER) run --rm -v $(shell pwd):/mnt $(PROTOC_IMAGE)
...
 
 # https://openapi-generator.tech
-OPENAPI_GENERATOR_IMAGE=openapitools/openapi-generator-cli:v5.3.0
+OPENAPI_GENERATOR_VERSION=v5.3.0
+OPENAPI_GENERATOR_IMAGE=openapitools/openapi-generator-cli:$(OPENAPI_GENERATOR_VERSION)
 OPENAPI_GENERATOR=$(DOCKER) run --user $(UID_GID) --rm -v $(shell pwd):/mnt $(OPENAPI_GENERATOR_IMAGE)
 
+OPENAPI_GENERATOR_VERSION_ADVANCED=v7.0.0
+OPENAPI_GENERATOR_IMAGE_ADVANCED=openapitools/openapi-generator-cli:$(OPENAPI_GENERATOR_VERSION_ADVANCED)
+OPENAPI_GENERATOR_ADVANCED=$(DOCKER) run --user $(UID_GID) --rm -v $(shell pwd):/mnt $(OPENAPI_GENERATOR_IMAGE_ADVANCED)
```
Comment (arielshaqed) [inline Makefile:21]: The more I think about these names the less I am a fan of them.  Perhaps call them "OPENAPI_GENERATOR_SERVER" and "OPENAPI_GENERATOR_CLIENT_SDK"?  (Extremely unimportant, any which way you decide)

File: Makefile [hunk 3]
```diff
@@ -108,8 +113,9 @@ tools: ## Install tools
 	$(GOCMD) install github.com/golangci/golangci-lint/cmd/golangci-lint@$(GOLANGCI_LINT_VERSION)
 	$(GOCMD) install google.golang.org/protobuf/cmd/protoc-gen-go
 
+client-python: python-client python-sdk
 
-client-python: api/swagger.yml  ## Generate SDK for Python client
+python-client: api/swagger.yml  ## Generate SDK for Python client - openapi generator version 5.3.0
```
Comment (arielshaqed) [inline Makefile:118]: Prefer to keep a common prefix for all SDKs, like we have common prefixes for other things here.
```suggestion
sdk-python-legacy: api/swagger.yml  ## Generate SDK for Python client - openapi generator version 5.3.0
```

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

- I would like to share a common prefix on all Makefile targets of SDKs / generated API
- I prefer to name the existing client targets "legacy"
- I think we need a check that the Python SDK is checked-in, like we already have for the "legacy" Python SDK
- Once ll this is done, please also check in the generated Python client.

File: clients/python/sdk/.openapi-generator-ignore [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+#!docs/README.md
+
+.gitlab-ci.yml
+git_push.sh
+.github/
+
+Gemfile.lock
+Gemfile
+_config.yml
+templates/
```
Comment (arielshaqed) [inline clients/python/sdk/.openapi-generator-ignore:32]: This file looks wrong.  Can we get a better file, maybe a "Python gitignore" will do better?

File: clients/python/templates/sdk/client.mustache [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                    base_path = parse_url(settings[0].get('url')).path
+                    configuration.host = Url(scheme=o.scheme, auth=o.auth, host=o.host, port=o.port,
+                                             path=base_path, query=o.query, fragment=o.fragment).url
+        except ValueError:
+            pass
+        return configuration
+
+    @property
+    def actions(self):
+        warnings.warn("Deprecated property. Use actions_api instead.", DeprecationWarning)
```
Comment (arielshaqed) [inline clients/python/templates/sdk/client.mustache:64]: It's a *new* SDK.  Let's not do this!

File: clients/python/templates/sdk/setup.mustache [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+# http://pypi.python.org/pypi/setuptools
+
+REQUIRES = [
+  "urllib3 >= 1.25.3",
+  "python-dateutil",
+{{#asyncio}}
+  "aiohttp >= 3.0.0",
+{{/asyncio}}
+{{#tornado}}
+  "tornado>=4.2,<5",
```
Comment (arielshaqed) [inline clients/python/templates/sdk/setup.mustache:24]: We need this?

File: clients/python/templates/sdk/setup.mustache [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    long_description = f.read()
+
+setup(
+    name=NAME,
+    version=VERSION,
+    description="{{appName}}",
+    author="{{infoName}}{{^infoName}}OpenAPI Generator community{{/infoName}}",
+    author_email="{{infoEmail}}{{^infoEmail}}team@openapitools.org{{/infoEmail}}",
+    url="{{packageUrl}}",
+    keywords=["OpenAPI", "OpenAPI-Generator", "{{{appName}}}"],
```
Comment (arielshaqed) [inline clients/python/templates/sdk/setup.mustache:42]:
```suggestion
    keywords=["OpenAPI", "auto-generated", "{{{appName}}}"],
```
It's _not_ an OpenAPI generator.  We can add something longer that says it was auto-generated in the description, though.

**Review by arielshaqed — COMMENTED** [review]
Comments about the template files, too!

File: Makefile (hunk 2 above)
Comment (Jonathan-Rosenberg) [reply Makefile:21]: the only difference between them is the openAPI version they use. This is the reason that I've added the `ADVANCED` part to the newer version. 
They both generate the client's SDK

File: .github/workflows/python-api-client-test-pypi.yaml [hunk 8]
```diff
@@ -0,0 +1,45 @@
+name: python-api-client-test
+
+on:
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: Tag/version to publish
+        required: true
+      project_name:
+        description: The Pypi (test) package name
```
Comment (arielshaqed) [inline .github/workflows/python-api-client-test-pypi.yaml:10]:
```suggestion
        description: The PyPI (test) package name
```

File: .github/workflows/python-api-client.yaml [hunk 9]
```diff
@@ -49,25 +57,44 @@ jobs:
...
+#    - name: Publish to docs repository
+#      uses: dmnemec/copy_file_to_another_repo_action@v1.1.1
+#      env:
+#        API_TOKEN_GITHUB: ${{ secrets.PERSONAL_TOKEN }}
+#      with:
+#        source_file: clients/python/_site/.
+#        destination_repo: treeverse/docs-lakefs-sdk-python
+#        destination_folder: /
+#        user_email: 'support@treeverse.io'
+#        user_name: 'python-docs-action'
```
Comment (arielshaqed) [inline .github/workflows/python-api-client.yaml:100]: ?

File: api/swagger2.yml [hunk 10]
```diff
@@ -0,0 +1,4631 @@
+openapi: "3.0.0"
```
Comment (arielshaqed) [inline api/swagger2.yml:1]: Do we need this file?

**Review by arielshaqed — APPROVED** [review]
Do we need swagger2.yml?  (Because it might be folded in the default GitHub view...)

Other than that -- small comments only.  Very very neat, thanks!

File: api/swagger2.yml (hunk 10 above)
Comment (Jonathan-Rosenberg) [reply api/swagger2.yml:1]: Good catch.
Was checked in by mistake

File: .github/workflows/python-api-client.yaml (hunk 9 above)
Comment (Jonathan-Rosenberg) [reply .github/workflows/python-api-client.yaml:100]: Disabled the creation of the new SDK until after I test it over TestPypi

**Pull request merged**